arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+  if (!paramsList.isArray() || paramsList.length() != 2) {
+    return EvalError("expecting to arguments, a function and a list");
+  }
+
+  auto functionSlice = paramsList.at(0);
+  auto list = paramsList.at(1);
+
+  if (list.isArray()) {
+    VPackArrayBuilder ab(&result);
+    for (VPackArrayIterator iter(list); iter.valid(); iter++) {
+      VPackBuilder parameter;
+      {
+        VPackArrayBuilder pb(&parameter);
+        parameter.add(VPackValue(iter.index()));
+        parameter.add(*iter);
+      }
+
+      auto res = EvaluateApply(ctx, functionSlice,
+                               VPackArrayIterator(parameter.slice()), result, false);
+      if (res.fail()) {
+        return res.error().wrapMessage("when mapping pair " + parameter.toJson());
+      }
+    }
+  } else if (list.isObject()) {
+    VPackObjectBuilder ob(&result);
+    for (VPackObjectIterator iter(list); iter.valid(); iter++) {
+      VPackBuilder parameter;
+      {
+        VPackArrayBuilder pb(&parameter);
+        parameter.add(iter.key());
+        parameter.add(iter.value());
+      }
+
+      VPackBuilder tempBuffer;
+
+      auto res = EvaluateApply(ctx, functionSlice,
+                               VPackArrayIterator(parameter.slice()), tempBuffer, false);
+      if (res.fail()) {
+        return res.error().wrapMessage("when mapping pair " + parameter.toJson());
+      }
+
+      result.add(iter.key());
+      result.add(tempBuffer.slice());
+    }
+  } else {
+    return EvalError("expected list or object, found: " + list.toJson());
+  }
+
+  return {};
+}
+
+EvalResult Prim_Reduce(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  if (!paramsList.isArray() || paramsList.length() < 2) {
+    return EvalError(
+        "expecting at least two arguments, a function and two dicts");
+  }
+
+  auto inputValue = paramsList.at(0);
+  auto functionSlice = paramsList.at(1);
+  auto inputAccumulator = paramsList.at(2);
+
+  if (inputAccumulator.isNone()) {
+    return EvalError("input accumulator is required but not set!");
+  }
+
+  // reduce(key, value, accumulator)
+  // iter can be either VPackArrayIterator or VPackObjectIterator
+  auto buildLambdaParameters = [&](VPackBuilder& parameter, auto& iter) {
+    {
+      VPackArrayBuilder pb(&parameter);
+      if constexpr (std::is_same_v<VPackObjectIterator&, decltype(iter)>) {
+        parameter.add(iter.key()); // object key
+      } else {
+        parameter.add(VPackValue(iter.index())); // array index
+      }
+
+      if (iter.isFirst()) {
+        parameter.add(iter.value());  // value
+        parameter.add(inputAccumulator);  // accum
+      } else {
+        parameter.add(iter.value());              // value
+        parameter.add(result.slice());  // accumulator / previous result
+      }
+    }
+  };
+
+  auto reduceArray = [&](auto& inputValue) -> EvalResult {
+    for (VPackArrayIterator iter(inputValue); iter.valid(); iter++) {
```

Comment:
VPackArrayIterator is not completely trivial. We should use preincrement on it.
```suggestion
    for (VPackArrayIterator iter(inputValue); iter.valid(); ++iter) {
```